component/loki/source/cloudflare/cloudflare.go

```diff
@@ -0,0 +1,144 @@
+package cloudflare
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/agent/component/common/loki/positions"
+	cft "github.com/grafana/agent/component/loki/source/cloudflare/internal/cloudflaretarget"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.cloudflare",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the
+// loki.source.cloudflare component.
+type Arguments struct {
+	Config    Config              `river:"config,block"`
+	ForwardTo []loki.LogsReceiver `river:"forward_to,attr"`
+}
+
+// Component implements the loki.source.cloudflare component.
+type Component struct {
+	opts    component.Options
+	metrics *cft.Metrics
+
+	mut    sync.RWMutex
+	fanout []loki.LogsReceiver
+	target *cft.Target
+
+	posFile positions.Positions
+	handler loki.LogsReceiver
+}
+
+// New creates a new loki.source.cloudflare component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	err := os.MkdirAll(o.DataPath, 0750)
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+	positionsFile, err := positions.New(o.Logger, positions.Config{
+		SyncPeriod:        10 * time.Second,
+		PositionsFile:     filepath.Join(o.DataPath, "positions.yml"),
+		IgnoreInvalidYaml: false,
+		ReadOnly:          false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:    o,
+		metrics: cft.NewMetrics(o.Registerer),
+		handler: make(loki.LogsReceiver),
+		fanout:  args.ForwardTo,
+		posFile: positionsFile,
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", "loki.source.cloudflare component shutting down, stopping the target")
+		c.target.Stop()
```

Comment:
There's a race condition here with `Update`; you need to grab the mutex before stopping the target.